server/src/main/java/org/apache/druid/discovery/LocalDruidNodeDiscoveryProvider.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.discovery;
+
+import org.apache.druid.server.DruidNode;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BooleanSupplier;
+
+/**
+ * Local node discovery. In this mode, Druid may run a number of
+ * logical servers within a single process, so that node discovery
+ * is all within the same process, but matches a true distributed
+ * node discovery.
+ * <p>
+ * Concurrency is at the level of the entire provider, which is
+ * necessary to coordinate the active state with ongoing node
+ * discovery and listener registrations. Such a "global" lock
+ * is fine because the actions don't occur frequently, nor do the
+ * actions take much time.
+ */
+public class LocalDruidNodeDiscoveryProvider extends DruidNodeDiscoveryProvider implements DruidNodeAnnouncer
+{
+  private class RoleEntry implements DruidNodeDiscovery
+  {
+    @SuppressWarnings("unused")
+    private final NodeRole role;
+    private final Map<DruidNode, DiscoveryDruidNode> nodes = new HashMap<>();
```

Comment:
I'm not 100% sure if this is valid. I think it's possible for there to be more than one DiscoveryDruidNode per DruidNode, if a node has more than one NodeRole. Although I haven't dug into this to be sure.

Reply:
Ah I believe you're right. The structure in this file is NodeRole -> DruidNode -> DruidNodeDiscovery, which does match how the ZK one works. When I was looking at it earlier, I missed that this map here is scoped to a single NodeRole.